allow regex on exludedPages

RobinGranstromKall opened the pull request:
Hi!

I have just started using this package in a project of mine. 
Discovered some issues with excluding pages and found a simple solution to my issue.

It's just making sure that all the items in the excludedPages list are regexes and does a test on those instead of a string match.
Previous syntax with string matching will work fine with these changes.

RobinGranstromKall committed:
allow regex on exludedPages

3nvi commented:
Hey! 

I don't mind allowing regexes there (although I'd honestly prefer globs over regexes). My issue has to do with the logic of this since certain characters need to be excluded when generating the regex. For example, take a look at this buggy behavior:

<img width="539" alt="Screenshot 2023-01-25 at 3 41 57 PM" src="https://user-images.githubusercontent.com/10436045/214578719-b00c0c4d-21fe-4e21-99cc-9f1f4488d920.png">

I think that this gets solved when movint to globs, but I'm open to sticking to regexes if we can safely convert strings to regexes :)

RobinGranstromKall commented:
Oh duh, that was a misstake 😅 Good catch!
Though I was under the assumption that this would take url paths to pages and not file names? But it's still an issue since the strings won't do what is expected.

I could investigate how to safely translate the flags in the regex, but how would you go about using globs? Perhaps that is the better way to do it?

3nvi commented:
> Oh duh, that was a misstake 😅 Good catch! Though I was under the assumption that this would take url paths to pages and not file names? But it's still an issue since the strings won't do what is expected.
> 
> I could investigate how to safely translate the flags in the regex, but how would you go about using globs? Perhaps that is the better way to do it?

You'd need to use a `blob` package for that to parse them. The approach is similar to what you have there though. The glob package would expose a way for you to test if a page matches a glob.

Alternatively, if you don't want to change the code here at all, you can use a regex in your project and using `fs` you can list all the paths that match your regex and just provide them to this plugin here

Commit:
feat: change regexp to glob via minimatch

RobinGranstromKall commented:
I adapted the implementation to use [minimatch](https://github.com/isaacs/minimatch) instead, it seems like it works as expected.

I wasn't sure how you want the dependencies to be listed, so i added minimatch in `gatsby-plugin-intl` as a peerDependency and in the root project i added it as a devDependency as it looked like you have done that for other peerDependencies, hopefully that was right.

I was contemplating weather to do with fs on my end, but what if there are pages created via gatsby-node that are not in the fs? So i thought this would cover all bases.
Or perhaps i'm misunderstanding gatsby and they are available to find via fs, i'm quite new when it comes to gatsby.

I also realised i managed to bypass your husky pre-commit checks on my first commit by using the GH web editor, my bad, if you want i can change that.

3nvi commented on packages/gatsby-plugin-intl/package.json:
```diff
@@ -14,6 +14,7 @@
   "peerDependencies": {
     "gatsby": ">=2.0.0",
     "gatsby-plugin-react-helmet": ">=3.1.6",
+    "minimatch": "^6.1.6",
```
I'm unsure as to whether this is indeed a `peerDependency`. In my eyes, this is an actual dependency since the runtime requires it.

3nvi commented on packages/gatsby-plugin-intl/README.md:
````diff
@@ -38,10 +38,10 @@ if you want a package that includes translations.
 
 ```
-npm i react-helmet gatsby-plugin-react-helmet @3nvi/gatsby-plugin-intl
+npm i react-helmet gatsby-plugin-react-helmet @3nvi/gatsby-plugin-intl minimatch
````
I don't think that this should be a peer dependency. Asking the user to install this doesn't make much sense 😄 We should list that as an actual dep of the package

3nvi commented on packages/gatsby-plugin-intl/src/gatsby-node.js:
```diff
@@ -22,10 +23,10 @@ export const onCreatePage = async (
   const isEnvDevelopment = process.env.NODE_ENV === 'development';
   const originalPath = page.path;
-  const is404 = originalPath.includes(notFoundPage);
+  const is404 = originalPath.includes(notFoundPage); // return early if page is excluded
```
could we move this comment down to L28?

3nvi commented on packages/gatsby-plugin-intl/src/gatsby-node.js:
```diff
@@ -4,6 +4,7 @@
  * it will be accessible from https:// .../en/404/ and https:// .../el/404/
  */
 import { DEFAULT_OPTIONS } from './constants';
+const minimatch = require('minimatch');
```
nit: can we flip the order of imports. Generally relative imports are last, but I seem to have forgotten this ESLint rule

RobinGranstromKall committed:
fix: change import order, move comment, change minimatch to dependency

RobinGranstromKall committed:
fix: update readme

RobinGranstromKall commented:
Hi! I looked over your comments and fixed the things you asked for! I re-requested review from you, but i know that that notification usually goes unnoticed.

RobinGranstromKall commented:
I have reconsidered my solution and i won't be needing this feature, so there is no stress from my part in getting this merged. 
But perhaps someone else might find it useful so i'll leave it up to you to decide if it's something that you think fits the plugin.
